### PR TITLE
avoid stopping rule groups if new rule groups are as same as old rule…

### PR DESCRIPTION
**Pull request by alburthoffman**
this is for the discussion in the user mail "[reload rule groups smartly](https://groups.google.com/d/msg/prometheus-users/zfqGXBQTelw/xq4udUhRBgAJ)".

basically enhance the rule groups reload process:

1. skip the old group if new group is equal to the old group
1. stop the old group and start the new group if existing and not equal
1. start the new added groups

The motivation for this is because we meet issues when reload rule groups.
The remove read endpoint is not stable. so sometimes it will take a long long time to stop all rule groups.
But actually we only need to update few rules like 1% of rules running.

### CODE REVIEW

File: rules/alerting.go [hunk 1]
```diff
@@ -538,3 +538,56 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+
+	if len(r.externalLabels) != len(ar.externalLabels) {
+		return false
+	}
+
+	same := true
+	for k, v := range r.externalLabels {
+		av, ok := ar.externalLabels[k]
+		if !ok {
+			same = false
```
Comment (roidelapluie) [inline rules/alerting.go]: return false

File: rules/alerting.go [hunk 2]
```diff
@@ -538,3 +538,56 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+
+	if r.annotations.Hash() != ar.annotations.Hash() {
+		return false
+	}
+
+	if len(r.externalLabels) != len(ar.externalLabels) {
+		return false
+	}
+
+	same := true
```
Comment (roidelapluie) [inline rules/alerting.go]: you do not need this

File: rules/alerting.go (hunk 1 above)
Comment (roidelapluie) [inline rules/alerting.go]: return false

File: rules/alerting.go [hunk 3]
```diff
@@ -538,3 +538,56 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+	for k, v := range r.externalLabels {
+		av, ok := ar.externalLabels[k]
+		if !ok {
+			same = false
+			break
+		}
+		if strings.Compare(v, av) != 0 {
+			same = false
```
Comment (roidelapluie) [inline rules/alerting.go]: return false

File: rules/alerting.go [hunk 4]
```diff
@@ -538,3 +538,56 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+			break
+		}
+	}
+
+	return same
```
Comment (roidelapluie) [inline rules/alerting.go]: return len(r.externalLabels) == len(ar.externalLabels)

File: rules/manager.go [hunk 5]
```diff
@@ -743,6 +747,53 @@ func (g *Group) RestoreForState(ts time.Time) {
...
+	for _, r := range g.rules {
+		gm[r.Name()] = r
+	}
+
+	ngm := make(map[string]Rule, len(ng.rules))
+	for _, r := range ng.rules {
+		ngm[r.Name()] = r
+	}
+
+	same := true
```
Comment (roidelapluie) [inline rules/manager.go]: you do not need this

File: rules/manager.go [hunk 6]
```diff
@@ -743,6 +747,53 @@ func (g *Group) RestoreForState(ts time.Time) {
...
+	for n, nr := range ngm {
+		or, ok := gm[n]
+		if !ok {
+			same = false
```
Comment (roidelapluie) [inline rules/manager.go]: return false

File: rules/manager.go [hunk 7]
```diff
@@ -743,6 +747,53 @@ func (g *Group) RestoreForState(ts time.Time) {
...
+			break
+		}
+		if !nr.Equals(or) {
+			same = false
```
Comment (roidelapluie) [inline rules/manager.go]: return false

File: rules/manager.go [hunk 8]
```diff
@@ -743,6 +747,53 @@ func (g *Group) RestoreForState(ts time.Time) {
...
+			break
+		}
+	}
+	return same
```
Comment (roidelapluie) [inline rules/manager.go]: return len(ngm) == len(gm)

File: rules/manager.go [hunk 9]
```diff
@@ -836,16 +887,21 @@ func (m *Manager) Update(interval time.Duration, files []string, externalLabels
 	m.restored = true
 
 	var wg sync.WaitGroup
-
+	fmt.Println(len(groups))
```
Comment (roidelapluie) [inline rules/manager.go]: debug thing?

Comment (alburthoffman) [reply rules/manager.go]: yeah. will remove it.

File: rules/alerting.go (hunk 2 above)
Comment (alburthoffman) [reply rules/alerting.go]: ok, this is better.

File: rules/alerting.go (hunk 4 above)
Comment (alburthoffman) [reply rules/alerting.go]: done

**Comment by alburthoffman** [comment]
@richardkiene please review

File: rules/alerting.go [hunk 10]
```diff
@@ -538,3 +538,49 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+
+	if strings.Compare(r.vector.String(), ar.vector.String()) != 0 {
+		return false
+	}
+
+	if r.holdDuration != ar.holdDuration {
+		return false
+	}
+
+	if r.labels.Hash() != ar.labels.Hash() {
```
Comment (codesome) [inline rules/alerting.go]: Using `labels.Equal` method would be faster. `!labels.Equal(r.labels, ar.labels)`. Same for annotations.

File: rules/alerting.go [hunk 11]
```diff
@@ -538,3 +538,49 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+		av, ok := ar.externalLabels[k]
+		if !ok {
+			return false
+		}
+		if strings.Compare(v, av) != 0 {
+			return false
+		}
+	}
+
+	return len(r.externalLabels) == len(ar.externalLabels)
```
Comment (codesome) [inline rules/alerting.go]: The length check should come before you check the map contents.

File: rules/manager.go [hunk 12]
```diff
@@ -743,6 +746,50 @@ func (g *Group) RestoreForState(ts time.Time) {
...
+
+	gm := make(map[string]Rule, len(g.rules))
+	for _, r := range g.rules {
+		gm[r.Name()] = r
+	}
+
+	ngm := make(map[string]Rule, len(ng.rules))
+	for _, r := range ng.rules {
+		ngm[r.Name()] = r
+	}
```
Comment (codesome) [inline rules/manager.go]: I am not sure if rule names can be repeated. Someone has to confirm.

File: rules/recording.go [hunk 13]
```diff
@@ -190,3 +191,28 @@ func (rule *RecordingRule) HTMLSnippet(pathPrefix string) template.HTML {
...
+		return false
+	}
+
+	if strings.Compare(r.vector.String(), rr.vector.String()) != 0 {
+		return false
+	}
+
+	if r.labels.Hash() != rr.labels.Hash() {
+		return false
+	}
```
Comment (codesome) [inline rules/recording.go]: Same comments that I had for `AlertingRule`

File: rules/alerting.go (hunk 10 above)
Comment (alburthoffman) [reply rules/alerting.go]: yeah. ur right. this is better.

File: rules/alerting.go [hunk 14]
```diff
@@ -538,3 +538,49 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+// 4. labels
+// 5. annotations
+// 6. externalLabels
+func (r *AlertingRule) Equals(rule Rule) bool {
+	ar, ok := rule.(*AlertingRule)
+	if !ok {
+		return false
+	}
+
+	if strings.Compare(r.name, ar.name) != 0 {
```
[the remark this answers is not in the document]
Comment (alburthoffman) [reply rules/alerting.go]: check the latest commit.

File: rules/recording.go (hunk 13 above)
Comment (alburthoffman) [reply rules/recording.go]: check the latest commit.

File: rules/alerting.go [hunk 15]
```diff
@@ -538,3 +538,53 @@ func (r *AlertingRule) HTMLSnippet(pathPrefix string) html_template.HTML {
...
+
+	if !labels.Equal(r.annotations, ar.annotations) {
+		return false
+	}
+
+	if len(r.externalLabels) != len(ar.externalLabels) {
+		return false
+	}
+
+	for k, v := range r.externalLabels {
```
Comment (roidelapluie) [inline rules/alerting.go]: if !labels.Equal(r.externalLabels, ar.externalLabels) {

would that work??

File: rules/manager.go [hunk 16]
```diff
@@ -24,7 +24,7 @@ import (
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
-	opentracing "github.com/opentracing/opentracing-go"
```
Comment (roidelapluie) [inline rules/manager.go]: unrelated?

File: rules/manager.go (hunk 12 above)
Comment (roidelapluie) [reply rules/manager.go]: yes you can have multiple rules with same name

File: rules/manager.go [hunk 17]
```diff
@@ -866,7 +916,6 @@ func (m *Manager) Update(interval time.Duration, files []string, externalLabels
 	for _, oldg := range m.groups {
 		oldg.stop()
 	}
-
```
Comment (roidelapluie) [inline rules/manager.go]: unrelated

File: rules/recording.go [hunk 18]
```diff
@@ -21,7 +21,7 @@ import (
 	"sync"
 	"time"
 
-	yaml "gopkg.in/yaml.v2"
```
Comment (roidelapluie) [inline rules/recording.go]: unrelated

File: rules/alerting.go (hunk 15 above)
Comment (codesome) [reply rules/alerting.go]: It is a `map[string]string` type

**Comment by roidelapluie** [comment]
The tests are probably too small to catch everything

**Comment by roidelapluie** [comment]
@alburthoffman What about instead of trying to compare logically, we would compare the YAML representations? That should catch about every change and would be simples.

File: rules/manager.go (hunk 16 above)
Comment (alburthoffman) [reply rules/manager.go]: not related. it is removed by `go mod tidy`

**Comment by alburthoffman** [comment]
@roidelapluie the problem is manager.Update does not have RuleGroups, which are parsed from yaml files directly. Also manager will have to store the old RuleGroups as well.

The Group struct and Rule struct in manager.Update does not have any yaml annotations yet.

File: rules/alerting.go (hunk 15 above)
Comment (alburthoffman) [reply rules/alerting.go]: yeah. it wont work

**Comment by johncming** [comment]
This comparison check  is  not faster than the original, if the update rule is more serious.

why not add Hash function in `AlertingRule` and `RecordingRule` which like `labels.Labels`.

https://github.com/prometheus/prometheus/blob/48d25e6fe7d960c76c01763303c38f11168b1bba/pkg/labels/labels.go#L121-L131

File: rules/manager.go (hunk 17 above)
Comment (alburthoffman) [reply rules/manager.go]: add it back

File: rules/manager.go (hunk 12 above)
Comment (alburthoffman) [reply rules/manager.go]: fix it by using Rule.String() as a key

File: rules/recording.go (hunk 18 above)
Comment (alburthoffman) [reply rules/recording.go]: roll it back

**Comment by alburthoffman** [comment]
@roidelapluie it's clean now. only manager.go gets modified. please check.

@johncming yeah. Hash will be much faster. but we will need to add Hash for AlertingRule and RecordingRule, Also probably Group as well. It's more easier to keep it simple, and performance might not be an issue here.

File: rules/manager.go [hunk 19]
```diff
@@ -743,6 +743,47 @@ func (g *Group) RestoreForState(ts time.Time) {
...
+		return false
+	}
+
+	if len(g.rules) != len(ng.rules) {
+		return false
+	}
+
+	gm := make(map[string]Rule, len(g.rules))
+	for _, r := range g.rules {
+		gm[r.String()] = r
```
Comment (roidelapluie) [inline rules/manager.go]: why not just using a slice? and them compare them one by once? changing the order of rules is a different group.

Comment (alburthoffman) [reply rules/manager.go]: So the order of the rules also matters, right?

Comment (alburthoffman) [reply rules/manager.go]: That would make things much easier

Comment (alburthoffman) [reply rules/manager.go]: done.

**Comment by alburthoffman** [comment]
@roidelapluie please review. thx

**Comment by johncming** [comment]
> @roidelapluie it's clean now. only manager.go gets modified. please check.
> 
> @johncming yeah. Hash will be much faster. but we will need to add Hash for AlertingRule and RecordingRule, Also probably Group as well. It's more easier to keep it simple, and performance might not be an issue here.

@alburthoffman 
 There is nothing wrong with the existing way, but for performance: let the unchanged old group keep running. If not for performance, then for what.

**Comment by alburthoffman** [comment]
"not an issue" means that performance might not be so bad that we have to optimize it at present. Here are our two choices for performance decision:

1. compare few fields (integers and strings)
2. calculate hash value and compare

not (stop all VS stop few). So I really do not thing the first choice is super slower than the second one. Honestly we need data for this kind of question, instead of assuming the second one is the best.

Remember that hash calculation is cpu intensive, and string comparison is simply looping...

**Comment by johncming** [comment]
> "not an issue" means that performance might not be so bad that we have to optimize it at present. Here are our two choices for performance decision:
> 
> 1. compare few fields (integers and strings)
> 2. calculate hash value and compare
> 
> not (stop all VS stop few). So I really do not thing the first choice is super slower than the second one. Honestly we need data for this kind of question, instead of assuming the second one is the best.
> 
> Remember that hash calculation is cpu intensive, and string comparison is simply looping...

For the fist method, fewer fields may affect comparison accuracy, which is unacceptable for rule files.

**Comment by alburthoffman** [comment]
In order to calculate the Hash value, you also have to decide which fields are used. Probably same fields as the ones in Equals method. So actually it makes no differences. 

I don't get ur second question. The changes aim to make less effort to archive the goal, like calling Rule.String() method instead of defining Equals in AlertingRule and RecordingRule.

So later, people who change String() wont need to change Equals() as well. Since no String() for Group, so we will have to add code to do it.

My point for performance is that don't rush to optimize it unless it is really needed. Normally we will need to profile and data for this.

**Comment by alburthoffman** [comment]
@roidelapluie please review.

**Comment by roidelapluie** [comment]
We have groups that take 2 minutes to run. I guess that this change will make it not needed to wait for them to stop.

**Comment by roidelapluie** [comment]
Can you add tests?

**Comment by brian-brazil** [comment]
I doubt rule groups will be more than a few MB, equals is fast enough for that. We can always optimise later.

**Comment by alburthoffman** [comment]
> Can you add tests?

sure. tests have been cleanup, so will add tests to manager_test.go

File: rules/manager.go [hunk 20]
```diff
@@ -743,6 +743,37 @@ func (g *Group) RestoreForState(ts time.Time) {
 
 }
 
+// Equals return if two groups have the same
+// 1. name
+// 2. file path
+// 3. interval
+// 4. rule set. and the rules are the same
```
Comment (roidelapluie) [inline rules/manager.go]: Maybe simplify thisEquals return if two groups are the same.

Comment (alburthoffman) [reply rules/manager.go]: that's ok. the code shows the same things. please check the latest commit .

**Comment by alburthoffman** [comment]
@roidelapluie please view.

**Comment by alburthoffman** [comment]
@brian-brazil are we good with this PR or still need enhancements? Thx

File: rules/manager.go [hunk 21]
```diff
@@ -836,16 +863,20 @@ func (m *Manager) Update(interval time.Duration, files []string, externalLabels
...
 	for _, newg := range groups {
-		wg.Add(1)
-
 		// If there is an old group with the same identifier, stop it and wait for
 		// it to finish the current iteration. Then copy it into the new group.
 		gn := groupKey(newg.name, newg.file)
 		oldg, ok := m.groups[gn]
 		delete(m.groups, gn)
 
+		// if equal, move it to groups
```
Comment (brian-brazil) [inline rules/manager.go]: Capital letter and full stop

File: rules/manager.go [hunk 22]
```diff
@@ -743,6 +743,33 @@ func (g *Group) RestoreForState(ts time.Time) {
 
 }
 
+// Equals return if two groups are the same
+func (g *Group) Equals(ng *Group) bool {
```
Comment (brian-brazil) [inline rules/manager.go]: This is going to be fragile when we add new fields, I'd suggest marshalling it and comparing that.

File: rules/manager.go [hunk 23]
```diff
@@ -836,16 +863,20 @@ func (m *Manager) Update(interval time.Duration, files []string, externalLabels
 	m.restored = true
 
 	var wg sync.WaitGroup
-
 	for _, newg := range groups {
-		wg.Add(1)
-
 		// If there is an old group with the same identifier, stop it and wait for
```
Comment (brian-brazil) [inline rules/manager.go]: Comment needs updating.

File: rules/manager_test.go [hunk 24]
```diff
@@ -703,18 +708,73 @@ func TestUpdate(t *testing.T) {
...
 	}
 
 	err = ruleManager.Update(10*time.Second, files, nil)
 	testutil.Ok(t, err)
-	for _, g := range ruleManager.groups {
+	for h, g := range ruleManager.groups {
 		for _, actual := range g.seriesInPreviousEval {
 			testutil.Equals(t, expected, actual)
 		}
+		// groups are the same because of no updates
```
Comment (brian-brazil) [inline rules/manager_test.go]: And for the rest of the comments

File: rules/manager.go (hunk 22 above)
Comment (roidelapluie) [reply rules/manager.go]: Marshalling would require public fields.

File: rules/manager.go (hunk 21 above)
Comment (alburthoffman) [reply rules/manager.go]: correct this.

File: rules/manager.go (hunk 23 above)
Comment (alburthoffman) [reply rules/manager.go]: correct this.

File: rules/manager_test.go (hunk 24 above)
Comment (alburthoffman) [reply rules/manager_test.go]: correct this.

File: rules/manager.go (hunk 22 above)
Comment (alburthoffman) [reply rules/manager.go]: there are two options to avoid define extra Equals method:
1. expose public fields and allow marshalling
2. enhance Hash function to include fields.

for #2, it's not a good choice, because the Hash function is used to calculate the evaluation offset.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by alburthoffman** [comment]
Thx
